src/System.Management.Automation/engine/AutomationEngine.cs

```diff
@@ -38,27 +38,19 @@ internal AutomationEngine(PSHost hostInterface, InitialSessionState iss)
 #if !UNIX
             // Update the env variable PathEXT to contain .CPL
             var pathext = Environment.GetEnvironmentVariable("PathEXT");
-            pathext = pathext ?? string.Empty;
-            bool cplExist = false;
-            if (pathext != string.Empty)
+
+            if (string.IsNullOrEmpty(pathext))
             {
-                string[] entries = pathext.Split(Utils.Separators.Semicolon);
-                foreach (string entry in entries)
-                {
-                    string ext = entry.Trim();
-                    if (ext.Equals(".CPL", StringComparison.OrdinalIgnoreCase))
-                    {
-                        cplExist = true;
-                        break;
-                    }
-                }
+                Environment.SetEnvironmentVariable("PathEXT", ".CPL");
             }
-
-            if (!cplExist)
+            else if (!(pathext.EndsWith(";.CPL", StringComparison.OrdinalIgnoreCase) ||
+                       pathext.StartsWith(".CPL;", StringComparison.OrdinalIgnoreCase) ||
+                       pathext.Contains(";.CPL;", StringComparison.OrdinalIgnoreCase) ||
+                       pathext.Equals(".CPL", StringComparison.OrdinalIgnoreCase)))
             {
-                pathext = (pathext == string.Empty) ? ".CPL" :
-                    pathext.EndsWith(";", StringComparison.OrdinalIgnoreCase)
-                    ? (pathext + ".CPL") : (pathext + ";.CPL");
+                // Fast skip if we already added the extention as ";.CPL".
+                // Fast skip if user already added the extention.
+                pathext = pathext + ";.CPL";
```

Comment:
There could be `;` at the end already, and in that case we shouldn't add `;.CPL`.

Reply:
Please see my previous comment.

Reply:
I think we should do it correctly. PowerShell doesn't have to change the existing value if it's ill-formed, but powershell should not generate ill-formed value when appending. It would be just as simple as `pathext += pathext[pathext.Length -1] == ';' ? ".CPL" : ";.CPL"`.

Reply:
If we get ".TXT;" it is already "ill-formed" and we don't corrupt it but we ignore the empty extension and process correctly. But what if user want the empty value for a custom scenario? :-) Unlike scenario...

Your fix added.